### PR TITLE
[data] Sanitize ExecutionResources

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

`ExecutionResources` serves as 2 different purposes: resource usage and resource limit. None values in it have different defaults (0 and inf) for those 2 use cases. This makes the class error prone, as users need to carefully handle None values. 

The best solution is to use 2 different classes for those 2 purposes. But it's already part of the public API. we don't want to break backward compatibility. 

This PR adds more wrapper methods to simplify the usage and avoid exposing None values to the external.  
This change will also remove many mypy typing hinting warnings of doing algebra operations between None and a float. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/execution/resource_manager.py [hunk 1]
```diff
@@ -195,7 +184,7 @@ def get_op_usage_str(self, op: PhysicalOperator) -> str:
                 f"op_outputs: {memory_string(self._mem_op_outputs[op])})"
             )
             if (
-                self.op_resource_allocator_enabled()
+                isinstance(self._op_resource_allocator, ReservationOpResourceAllocator)
```
Comment (bveeramani) [inline python/ray/data/_internal/execution/resource_manager.py:187]: What's the motivation behind this change?

File: python/ray/data/_internal/execution/interfaces/execution_options.py [hunk 2]
```diff
@@ -183,10 +223,12 @@ class ExecutionOptions:
             option is useful for performance debugging. On by default.
     """
 
-    resource_limits: ExecutionResources = field(default_factory=ExecutionResources)
+    _resource_limits: ExecutionResources = field(
```
Comment (bveeramani) [inline python/ray/data/_internal/execution/interfaces/execution_options.py:226]: Do we need to worry about users (or us) specifying `resource_limits` in the `ExecutionOptions` constructor? Currently, it'll error.

```python
>>> ExecutionOptions(resource_limits=...)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ExecutionOptions.__init__() got an unexpected keyword argument 'resource_limits'
```

File: python/ray/data/_internal/execution/interfaces/execution_options.py [hunk 3]
```diff
@@ -7,149 +7,189 @@
...
     """
 
-    # CPU usage in cores (Ray logical CPU slots).
-    cpu: Optional[float] = None
+    def __init__(
+        self,
+        cpu: Optional[float] = None,
+        gpu: Optional[float] = None,
+        object_store_memory: Optional[float] = None,
+        default_to_inf: bool = False,
```
Comment (bveeramani) [inline python/ray/data/_internal/execution/interfaces/execution_options.py:22]: I'm wondering about how we set `default_to_inf` to `False` by default.

If you previously did something like this, your code would break, right? Because then you're specifying all-zero resource limits?

```python
DataContext.get_current().execution_options.resource_limits = ExecutionResources()
```

File: python/ray/data/_internal/execution/resource_manager.py (hunk 1 above)
Comment (raulchen) [reply python/ray/data/_internal/execution/resource_manager.py:187]: oh, I was cleaning up the type hint errors in this file, and noticed this.
If not changing this, mypy will complain on the next line.

File: python/ray/data/_internal/execution/interfaces/execution_options.py (hunk 3 above)
Comment (raulchen) [reply python/ray/data/_internal/execution/interfaces/execution_options.py:22]: This won't break because we've overridden the setter method for `resource_limits`.
And the reason of defaulting to False is because there are fewer places that should default to False.

File: python/ray/data/_internal/execution/interfaces/execution_options.py (hunk 2 above)
Comment (raulchen) [reply python/ray/data/_internal/execution/interfaces/execution_options.py:226]: Good point. there are indeed some test failures because of this.

Comment (raulchen) [reply python/ray/data/_internal/execution/interfaces/execution_options.py]: fixed

**Review by bveeramani — APPROVED** [review]
